semver/parse.ts

```diff
@@ -0,0 +1,635 @@
+  ")*))";
+
+// ## Build Metadata Identifier
+// Any combination of digits, letters, or hyphens.
+
+const BUILDIDENTIFIER: number = R++;
+src[BUILDIDENTIFIER] = "[0-9A-Za-z-]+";
+
+// ## Build Metadata
+// Plus sign, followed by one or more period-separated build metadata
+// identifiers.
+
+const BUILD: number = R++;
+src[BUILD] = "(?:\\+(" + src[BUILDIDENTIFIER] + "(?:\\." +
+  src[BUILDIDENTIFIER] + ")*))";
+
+// ## Full Version String
+// A main version, followed optionally by a pre-release version and
+// build metadata.
+
+// Note that the only major, minor, patch, and pre-release sections of
+// the version string are capturing groups.  The build metadata is not a
+// capturing group, because it should not ever be used in version
+// comparison.
+
+const FULL: number = R++;
+const FULLPLAIN = "v?" + src[MAINVERSION] + src[PRERELEASE] + "?" + src[BUILD] +
+  "?";
+
+src[FULL] = "^" + FULLPLAIN + "$";
+
+const GTLT: number = R++;
+src[GTLT] = "((?:<|>)?=?)";
+
+// Something like "2.*" or "1.2.x".
+// Note that "x.x" is a valid xRange identifer, meaning "any version"
+// Only the first item is strictly required.
+const XRANGEIDENTIFIER: number = R++;
+src[XRANGEIDENTIFIER] = src[NUMERICIDENTIFIER] + "|x|X|\\*";
+
+const XRANGEPLAIN: number = R++;
+src[XRANGEPLAIN] = "[v=\\s]*(" +
+  src[XRANGEIDENTIFIER] +
+  ")" +
+  "(?:\\.(" +
+  src[XRANGEIDENTIFIER] +
+  ")" +
+  "(?:\\.(" +
+  src[XRANGEIDENTIFIER] +
+  ")" +
+  "(?:" +
+  src[PRERELEASE] +
+  ")?" +
+  src[BUILD] +
+  "?" +
+  ")?)?";
+
+const XRANGE: number = R++;
+src[XRANGE] = "^" + src[GTLT] + "\\s*" + src[XRANGEPLAIN] + "$";
+
+// Tilde ranges.
+// Meaning is "reasonably at or greater than"
+const LONETILDE: number = R++;
+src[LONETILDE] = "(?:~>?)";
+
+const TILDE: number = R++;
+src[TILDE] = "^" + src[LONETILDE] + src[XRANGEPLAIN] + "$";
+
+// Caret ranges.
+// Meaning is "at least and backwards compatible with"
+const LONECARET: number = R++;
+src[LONECARET] = "(?:\\^)";
+
+const CARET: number = R++;
+src[CARET] = "^" + src[LONECARET] + src[XRANGEPLAIN] + "$";
+
+// A simple gt/lt/eq thing, or just "" to indicate "any version"
+const COMPARATOR: number = R++;
+src[COMPARATOR] = "^" + src[GTLT] + "\\s*(" + FULLPLAIN + ")$|^$";
+
+// Something like `1.2.3 - 1.2.4`
+const HYPHENRANGE: number = R++;
+src[HYPHENRANGE] = "^\\s*(" +
+  src[XRANGEPLAIN] +
+  ")" +
+  "\\s+-\\s+" +
+  "(" +
+  src[XRANGEPLAIN] +
+  ")" +
+  "\\s*$";
+
+// Star ranges basically just allow anything at all.
+const STAR: number = R++;
+src[STAR] = "(<|>)?=?\\s*\\*";
+
+// Compile to actual regexp objects.
+// All are flag-free, unless they were created above with a flag.
+for (let i = 0; i < R; i++) {
+  if (!re[i]) {
+    re[i] = new RegExp(src[i]);
+  }
+}
+
+/**
+ * Returns the parsed version, or undefined if it's not valid.
+ * @param version The version string to parse
+ * @returns A valid SemVer or `undefined`
+ */
+export function tryParse(version?: string): SemVer | undefined {
```

Comment:
I think if you alter the version to be `version?: string | SemVer` that may reduce some dulicate code elsewhere which is attempting to differentiate on inputs. Also it would be consistent with the parse function below.